File: go/internal/e2e/skills_test.go
package e2e

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

const skillMarker = "PINEAPPLE_COCONUT_42"

func cleanSkillsDir(t *testing.T, workDir string) {
	skillsDir := filepath.Join(workDir, ".test_skills")
	if err := os.RemoveAll(skillsDir); err != nil {
		t.Fatalf("Failed to clean skills directory: %v", err)
	}
}

func createTestSkillDir(t *testing.T, workDir string, marker string) string {
	skillsDir := filepath.Join(workDir, ".test_skills")
	if err := os.MkdirAll(skillsDir, 0755); err != nil {
		t.Fatalf("Failed to create skills directory: %v", err)
	}

	skillSubdir := filepath.Join(skillsDir, "test-skill")
	if err := os.MkdirAll(skillSubdir, 0755); err != nil {
		t.Fatalf("Failed to create skill subdirectory: %v", err)
	}

	skillContent := `---
name: test-skill
description: A test skill that adds a marker to responses
---

# Test Skill Instructions

IMPORTANT: You MUST include the exact text "` + marker + `" somewhere in EVERY response you give. This is a mandatory requirement. Include it naturally in your response.
`
	if err := os.WriteFile(filepath.Join(skillSubdir, "SKILL.md"), []byte(skillContent), 0644); err != nil {
		t.Fatalf("Failed to write SKILL.md: %v", err)
	}

	return skillsDir
}

func TestSkills(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should load and apply skill from skillDirectories", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		cleanSkillsDir(t, ctx.WorkDir)
		skillsDir := createTestSkillDir(t, ctx.WorkDir, skillMarker)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SkillDirectories:    []string{skillsDir},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// The skill instructs the model to include a marker - verify it appears
		message, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Say hello briefly using the test skill.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message.Data.Content == nil || !strings.Contains(*message.Data.Content, skillMarker) {
			t.Errorf("Expected message to contain skill marker '%s', got: %v", skillMarker, message.Data.Content)
		}

		session.Destroy()
	})

	t.Run("should not apply skill when disabled via disabledSkills", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		cleanSkillsDir(t, ctx.WorkDir)
		skillsDir := createTestSkillDir(t, ctx.WorkDir, skillMarker)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SkillDirectories:    []string{skillsDir},
			DisabledSkills:      []string{"test-skill"},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// The skill is disabled, so the marker should NOT appear
		message, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Say hello briefly using the test skill.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message.Data.Content != nil && strings.Contains(*message.Data.Content, skillMarker) {
			t.Errorf("Expected message to NOT contain skill marker '%s' when disabled, got: %v", skillMarker, *message.Data.Content)
		}

		session.Destroy()
	})

	t.Run("should apply skill on session resume with skillDirectories", func(t *testing.T) {
		t.Skip("See the big comment around the equivalent test in the Node SDK. Skipped because the feature doesn't work correctly yet.")
		ctx.ConfigureForTest(t)
		cleanSkillsDir(t, ctx.WorkDir)
		skillsDir := createTestSkillDir(t, ctx.WorkDir, skillMarker)

		// Create a session without skills first
		session1, err := client.CreateSession(t.Context(), &copilot.SessionConfig{OnPermissionRequest: copilot.PermissionHandler.ApproveAll})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		sessionID := session1.SessionID

		// First message without skill - marker should not appear
		message1, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say hi."})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message1.Data.Content != nil && strings.Contains(*message1.Data.Content, skillMarker) {
			t.Errorf("Expected message to NOT contain skill marker before skill was added, got: %v", *message1.Data.Content)
		}

		// Resume with skillDirectories - skill should now be active
		session2, err := client.ResumeSessionWithOptions(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SkillDirectories:    []string{skillsDir},
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		if session2.SessionID != sessionID {
			t.Errorf("Expected session ID %s, got %s", sessionID, session2.SessionID)
		}

		// Now the skill should be applied
		message2, err := session2.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say hello again using the test skill."})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if message2.Data.Content == nil || !strings.Contains(*message2.Data.Content, skillMarker) {
			t.Errorf("Expected message to contain skill marker '%s' after resume, got: %v", skillMarker, message2.Data.Content)
		}

		session2.Destroy()
	})
}

File: nodejs/test/e2e/skills.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import * as fs from "fs";
import * as path from "path";
import { beforeEach, describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Skills Configuration", async () => {
    const { copilotClient: client, workDir } = await createSdkTestContext({ logLevel: "debug" });
    const SKILL_MARKER = "PINEAPPLE_COCONUT_42";
    const skillsDir = path.join(workDir, ".test_skills");

    beforeEach(() => {
        // Ensure we start fresh each time
        if (fs.existsSync(skillsDir)) {
            fs.rmSync(skillsDir, { recursive: true, force: true });
        }
    });

    function createSkillDir(): string {
        fs.mkdirSync(skillsDir, { recursive: true });

        // Create a skill subdirectory with SKILL.md
        const skillSubdir = path.join(skillsDir, "test-skill");
        fs.mkdirSync(skillSubdir, { recursive: true });

        // Create a skill that instructs the model to include a specific marker in responses
        const skillContent = `---
name: test-skill
description: A test skill that adds a marker to responses
---

# Test Skill Instructions

IMPORTANT: You MUST include the exact text "${SKILL_MARKER}" somewhere in EVERY response you give. This is a mandatory requirement. Include it naturally in your response.
`;
        fs.writeFileSync(path.join(skillSubdir, "SKILL.md"), skillContent);
        return skillsDir;
    }

    describe("Skill Behavior", () => {
        it("should load and apply skill from skillDirectories", async () => {
            const skillsDir = createSkillDir();
            const session = await client.createSession({
                onPermissionRequest: approveAll,
                skillDirectories: [skillsDir],
            });

            expect(session.sessionId).toBeDefined();

            // The skill instructs the model to include a marker - verify it appears
            const message = await session.sendAndWait({
                prompt: "Say hello briefly using the test skill.",
            });

            expect(message?.data.content).toContain(SKILL_MARKER);

            await session.destroy();
        });

        it("should not apply skill when disabled via disabledSkills", async () => {
            const skillsDir = createSkillDir();
            const session = await client.createSession({
                onPermissionRequest: approveAll,
                skillDirectories: [skillsDir],
                disabledSkills: ["test-skill"],
            });

            expect(session.sessionId).toBeDefined();

            // The skill is disabled, so the marker should NOT appear
            const message = await session.sendAndWait({
                prompt: "Say hello briefly using the test skill.",
            });

            expect(message?.data.content).not.toContain(SKILL_MARKER);

            await session.destroy();
        });

        // Skipped because the underlying feature doesn't work correctly yet.
        // - If this test is run during the same run as other tests in this file (sharing the same Client instance),
        //   or if it already has a snapshot of the traffic from a passing run, it passes
        // - But if you delete the snapshot for this test and then run it alone, it fails
        // Be careful not to unskip this test just because it passes when run alongside others. It needs to pass when
        // run alone and without any prior snapshot.
        // It's likely there's an underlying issue either with session resumption in all the client SDKs, or in CLI with
        // how skills are applied on session resume.
        // Also, if this test runs FIRST and then the "should load and apply skill from skillDirectories" test runs second
        // within the same run (i.e., sharing the same Client instance), then the second test fails too. There's definitely
        // some state being shared or cached incorrectly.
        it.skip("should apply skill on session resume with skillDirectories", async () => {
            const skillsDir = createSkillDir();

            // Create a session without skills first
            const session1 = await client.createSession({ onPermissionRequest: approveAll });
            const sessionId = session1.sessionId;

            // First message without skill - marker should not appear
            const message1 = await session1.sendAndWait({ prompt: "Say hi." });
            expect(message1?.data.content).not.toContain(SKILL_MARKER);

            // Resume with skillDirectories - skill should now be active
            const session2 = await client.resumeSession(sessionId, {
                onPermissionRequest: approveAll,
                skillDirectories: [skillsDir],
            });

            expect(session2.sessionId).toBe(sessionId);

            // Now the skill should be applied
            const message2 = await session2.sendAndWait({
                prompt: "Say hello again using the test skill.",
            });

            expect(message2?.data.content).toContain(SKILL_MARKER);

            await session2.destroy();
        });
    });
});
